### PR TITLE
Dedup and refactor image acquisition

### PR DESCRIPTION
**Pull request by baude**
As promised in #19596, this pr deduplicates and refactors image acquisition.  All virt providers that use FCOS as its default now use the same code.

[NO NEW TESTS NEEDED]

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: pkg/machine/fcos.go [hunk 1]
```diff
@@ -107,37 +107,50 @@ type FcosDownload struct {
...
+		CacheDir:        cacheDir,
+		CompressionType: vp.Compression(),
+		DataDir:         dataDir,
+		Format:          vp.Format(),
+		VMKind:          vp.VMType(),
+		VMName:          vmName,
+	}, nil
+}
+
+func NewFcosDownloader(dl Download, imageStream FCOSStream) (DistributionDownload, error) {
```
Comment (vrothberg) [inline pkg/machine/fcos.go:132]: This should be turned into a method of `Download`

```golang
func (dl *Download) NewFcosDownloader(...)
```

Reasoning is that it's changing the passed `dl` and a method is more expressive than a somehow hidden in/out parameter.

**Review by vrothberg — COMMENTED** [review]
One nit.  Did not test but code LGTM

File: pkg/machine/fcos.go [hunk 2]
```diff
@@ -107,37 +107,50 @@ type FcosDownload struct {
 	Download
 }
 
-func NewFcosDownloader(vmType VMType, vmName string, imageStream FCOSStream, vp VirtProvider) (DistributionDownload, error) {
-	info, err := GetFCOSDownload(vp, imageStream)
+func NewDownload(vmName string, vp VirtProvider) (Download, error) {
```
Comment (rhatdan) [inline pkg/machine/fcos.go:110]: Should this be a VirtProvider function?

func (vp *VirtProfider) NewDownload(vmName string) (Download, error)

File: pkg/machine/fcos.go [hunk 3]
```diff
@@ -344,7 +366,53 @@ func IsValidFCOSStreamString(s string) bool {
...
+		return true
 	}
 
 	return false
 }
+
+// AcquireVMImage determines if the image is already in a FCOS stream. If so,
+// retrieves the image path of the uncompressed file. Otherwise, the user has
+// provided an alternative image, so we set the image path and download the image.
+func AcquireVMImage(dl Download, imagePath string) (*VMFile, FCOSStream, error) {
```
Comment (rhatdan) [inline pkg/machine/fcos.go:379]: Download function?

**Review by rhatdan — APPROVED** [review]

File: pkg/machine/fcos.go (hunk 2 above)
Comment (mheon) [reply pkg/machine/fcos.go:110]: Also, should probably return *Download - so you don't have to return Download{} but can return nil instead on error

File: pkg/machine/fcos.go [hunk 4]
```diff
@@ -216,9 +229,8 @@ func getStreamURL(streamType FCOSStream) url2.URL {
 	return fedoracoreos.GetStreamURL(streamType.String())
 }
 
-// This should get Exported and stay put as it will apply to all fcos downloads
-// getFCOS parses fedoraCoreOS's stream and returns the image download URL and the release version
-func GetFCOSDownload(vp VirtProvider, imageStream FCOSStream) (*FcosDownloadInfo, error) {
+// GetFCOSDownload parses fedoraCoreOS's stream and returns the image download URL and the release version
+func GetFCOSDownload(dl *Download, imageStream FCOSStream) (*FcosDownloadInfo, error) {
```
Comment (rhatdan) [inline pkg/machine/fcos.go:233]: func (dl *Download) FCOSDownload(imageStream FCOSStream) (*FcosDownloadInfo, error) {

File: pkg/machine/pull.go [hunk 5]
```diff
@@ -388,8 +388,8 @@ func RemoveImageAfterExpire(dir string, expire time.Duration) error {
 
 // AcquireAlternateImage downloads the alternate image the user provided, which
 // can be a file path or URL
-func AcquireAlternateImage(name string, vmtype VMType, opts InitOptions) (*VMFile, error) {
-	g, err := NewGenericDownloader(vmtype, name, opts.ImagePath)
+func AcquireAlternateImage(name string, vmtype VMType, inputPath string) (*VMFile, error) {
```
Comment (rhatdan) [inline pkg/machine/pull.go:391]: func (vmtype *VMType)  AcquireAlternateImage...

File: pkg/machine/config.go [hunk 6]
```diff
@@ -493,3 +524,76 @@ func WaitAndPingAPI(sock string) {
...
+	}
+	urlSplit := strings.Split(info.Location, "/")
+	dl.ImageName = urlSplit[len(urlSplit)-1]
+	downloadURL, err := url.Parse(info.Location)
+	if err != nil {
+		return nil, err
+	}
+
+	// Complete the download struct
+	dl.Arch = GetFcosArch()
```
Comment (vrothberg) [inline pkg/machine/config.go:541]: To be sure: are any of these assignments expected to be permanent and extend the scope of the call?

If the assignments are to be made permanent, the receiver `(dl Download)` must be changed to a pointer receiver `(dl *Download)`.

**Review by vrothberg — APPROVED** [review]
LGTM
Just one Q

File: pkg/machine/config.go (hunk 6 above)
Comment (baude) [reply pkg/machine/config.go:541]: I *think* it is OK as is.

**Pull request merged**